test/uk/gov/hmrc/agentsubscription/audit/AuditServiceSpec.scala

```diff
@@ -0,0 +1,135 @@
+
+import org.mockito.ArgumentCaptor
+import org.mockito.ArgumentMatchers.any
+import org.mockito.Mockito.{verify, when}
+import org.scalatest.concurrent.Eventually
+import org.scalatest.mock.MockitoSugar
+import play.api.libs.json.Json
+import play.api.test.FakeRequest
+import uk.gov.hmrc.agentsubscription.audit.AgentSubscriptionEvent.AgentSubscription
+import uk.gov.hmrc.play.audit.http.connector.{AuditConnector, AuditResult}
+import uk.gov.hmrc.play.audit.model.{AuditEvent, ExtendedDataEvent}
+import uk.gov.hmrc.play.http.HeaderCarrier
+import uk.gov.hmrc.play.http.logging.{Authorization, RequestId, SessionId}
+import uk.gov.hmrc.play.test.UnitSpec
+
+import scala.concurrent.{ExecutionContext, Future}
+
+class AuditServiceSpec extends UnitSpec with MockitoSugar with Eventually {
+  "auditEvent" should {
+    "send an event with the correct fields" in {
+      val mockConnector = mock[AuditConnector]
+      when(mockConnector.sendEvent(any[AuditEvent])(any[HeaderCarrier], any[ExecutionContext])).thenReturn(Future successful AuditResult.Success)
+      val service = new AuditService(mockConnector)
+
+      val hc = HeaderCarrier(
+        authorization = Some(Authorization("dummy bearer token")),
+        sessionId = Some(SessionId("dummy session id")),
+        requestId = Some(RequestId("dummy request id")),
+        trueClientIp = Some("1.1.1.1"),
+        trueClientPort = Some("12345")
+      )
+
+      val detail = Json.obj(
+        "agencyName" -> "Test Agency",
+        "agencyAddress" -> Json.obj(
+          "addressLine1" -> "1 Test Street",
+          "addressLine2" -> "Test village"
+        )
+      )
+      service.auditEvent(
+        AgentSubscription,
+        "transaction name",
+        detail
+      )(
+        hc,
+        FakeRequest("GET", "/path")
+      )
+
+      eventually {
+        val captor = ArgumentCaptor.forClass(classOf[AuditEvent])
+        verify(mockConnector).sendEvent(captor.capture())(any[HeaderCarrier], any[ExecutionContext])
+        captor.getValue shouldBe an[ExtendedDataEvent]
+        val sentEvent = captor.getValue.asInstanceOf[ExtendedDataEvent]
+
+        sentEvent.auditSource shouldBe "agent-subscription"
+        sentEvent.auditType shouldBe "AgentSubscription"
+
+        (sentEvent.detail \ "Authorization").as[String] shouldBe "dummy bearer token"
+
+        (sentEvent.detail \ "agencyName").as[String] shouldBe "Test Agency"
+        (sentEvent.detail \ "agencyAddress" \ "addressLine1").as[String] shouldBe "1 Test Street"
+        (sentEvent.detail \ "agencyAddress" \ "addressLine2").as[String] shouldBe "Test village"
+
+        sentEvent.tags("transactionName") shouldBe "transaction name"
+        sentEvent.tags("path") shouldBe "/path"
+        sentEvent.tags("X-Session-ID") shouldBe "dummy session id"
+        sentEvent.tags("X-Request-ID") shouldBe "dummy request id"
+
+        sentEvent.tags("clientIP") shouldBe "1.1.1.1"
+        sentEvent.tags("clientPort") shouldBe "12345"
```

Comment:
AuditService is generic so this test doesn't test the specifics of the AgentSubscription audit event, it just includes enough of the address to test that AuditService works when given nested fields (`"agencyAddress" \ "addressLine1"`).

The full set of AgentSubscription event details is tested in SubscriptionServiceSpec.